core/embed/rust/src/trezorhal/ble/mod.rs

```diff
@@ -0,0 +1,163 @@
+#[cfg(feature = "micropython")]
+mod micropython;
+
+use crate::error::Error;
+use core::mem::size_of;
+
+use super::{ffi, model};
+
+pub const ADV_NAME_LEN: usize = ffi::BLE_ADV_NAME_LEN as usize;
+pub const PAIRING_CODE_LEN: usize = ffi::BLE_PAIRING_CODE_LEN as usize;
+pub const RX_PACKET_SIZE: usize = ffi::BLE_RX_PACKET_SIZE as usize;
+pub const TX_PACKET_SIZE: usize = ffi::BLE_TX_PACKET_SIZE as usize;
+
+const OPERATION_FAILED: Error = Error::RuntimeError(c"BLE operation failed");
+
+// NOTE: replace with floor_char_boundary when stable
+fn prefix_utf8_bytes(text: &str, max_len: usize) -> &[u8] {
+    let mut i = text.len().min(max_len);
+    while !text.is_char_boundary(i) {
+        i -= 1;
+    }
+    &text.as_bytes()[..i]
+}
+
+fn state() -> ffi::ble_state_t {
+    let mut state = ffi::ble_state_t {
+        connected: false,
+        peer_count: 0,
+        connectable: false,
+        pairing: false,
+        pairing_requested: false,
+        state_known: false,
+    };
+    unsafe { ffi::ble_get_state(&mut state as _) };
+    state
+}
+
+fn issue_command(
+    cmd_type: ffi::ble_command_type_t,
+    cmd_data: ffi::ble_command_data_t,
+) -> Result<(), Error> {
+    let data_len = match cmd_type {
+        ffi::ble_command_type_t_BLE_ALLOW_PAIRING => PAIRING_CODE_LEN,
+        ffi::ble_command_type_t_BLE_PAIRING_MODE | ffi::ble_command_type_t_BLE_SWITCH_ON => {
+            size_of::<ffi::ble_adv_start_cmd_data_t>()
+        }
+        _ => 0,
+    };
+    let mut cmd = ffi::ble_command_t {
+        cmd_type,
+        data_len: unwrap!(data_len.try_into()),
+        data: cmd_data,
+    };
+    if unsafe { ffi::ble_issue_command(&mut cmd as _) } {
+        Ok(())
+    } else {
+        Err(OPERATION_FAILED)
+    }
+}
+
+fn data_advname(name: &str) -> ffi::ble_command_data_t {
+    let mut data = ffi::ble_command_data_t {
+        adv_start: ffi::ble_adv_start_cmd_data_t {
+            name: [0u8; ADV_NAME_LEN],
+            static_mac: false,
+        },
+    };
+    let bytes = prefix_utf8_bytes(name, ADV_NAME_LEN);
```

Comment:
Maybe it would be simpler to return an error when `name` is too long?

Reply:
Currently we are using the device label (max 32 chars) as the advertising name, so we need to be able to handle when it's longer than `ADV_NAME_LEN` (20). We can implement this in python instead but not sure if we can do anything better than using a prefix.